rfcs/0027-frontier-synchronization.md

```diff
@@ -0,0 +1,188 @@
+
+IMPLEMENTATION NOTE: The entire chain of mask destruction hooks needs to be executed synchronously within a single async cycle in order to update breadcrumb subtrees without fear of race conditions.
+
+#### Frontier Read Monad
+
+A new monad is introduced for specifying reads from the transition frontier (see [alternatives section](#alternatives-misread) for explanation on why). All existing read functions on the transition frontier will instead by turned into functions which do not take in a transition frontier and return a result wrapped in the `'a Transition_frontier.Read.t` monad. The monad is designed to interact with deferreds so that async programming is still accessible during reads when necessary.
+
+#### New Frontier Interface
+
+Below is a partial signature for what the new transition frontier interface would look like. Note that all base read/write functions from the transition frontier are removed, and instead, only the read monad and write actions are accessible. The only way to interact with the transition frontier thus becomes the `read` and `write` functions.
+
+```ocaml
+module Transition_frontier : sig
+  module Breadcrumb : sig
+    module Full : sig
+      type t
+      val block : t -> Block.Validated.t
+      val staged_ledger : t -> Transaction_snark_scan_state.t
+    end
+
+    module Partial : sig
+      type t
+      val block : t -> Block.Validated.t
+      val scan_state : t -> Transaction_snark_scan_state.t
+    end
+
+    type full
+    type partial
+
+    type 'typ t =
+      | Full : Full.t -> full t
+      | Partial : Partial.t -> partial t
+    type 'typ breadcrumb = 'typ t
+
+    val external_transition : t -> External_transition.Validated.t
+
+    module Any : sig
+      type t = T : _ breadcrumb -> t
+      type any = t
+
+      val wrap : _ breadcrumb -> t
+
+      module Ref : sig
+        (* intentionally left abstract to enforce synchronous access *)
+        type t
+
+        val wrap : any -> t
+        (* downgrades a full to a partial; returns error if already a partial *)
+        val downgrade : t -> unit Or_error.t
+
+        val external_transition : t -> External_transition.Validated.t
+        val staged_state :
+             t
+          -> [ `Staged_ledger of Staged_ledger.t
+             | `Scan_state of Transaction_snark_scan_state.t ]
+      end
+    end
+  end
+
+  module Extensions : sig
+    type t
+
+    (* ... *)
+  end
+
+  module Read : sig
+    include Monad.S
+
+    (* to wrap computations in the monad *)
+    val deferred : 'a Deferred.t -> 'a t
```

Comment:
nit: `lift_deferred`

Reply:
I think `deferred` reads better personally, but I can update it to follow the pattern of monads that embed other monads in our codebase.